### PR TITLE
Add advanced “TEMPLATE_ENGINE.md” documentation.

### LINKED ISSUE #187: Consolidate / improve internal template engine logic.
Because the entirety of the templating engine is opaque to integrators, the internal interfaces have never really been scrutinized.

Fortunately, _because_ it’s all internal, we can take this step-by-step and incrementally improve upon what we have without needing to do major / minor version bumps.

Some things to TODO:
* Improve class / method names to match expectations.
* Reduce the number of _layers_ (ideally, we eventually just have a `TemplateEngine` class).
* Create some documentation describing how things work at a high-level.

### PR DESCRIPTION
**Pull request by theengineear**
The point of this document is to explain (to folks that are already experts in web development) how html template engines conceptually work.

It walks the reader through a series of iterations which add increasing sophistication to a toy template engine library.

Closes #187.

### CODE REVIEW

**Comment by theengineear** [comment]
@klebba — Took some time (maybe more than I’d like to admit) trying to update these docs as you requested.

As discussed, these are aimed at folks who are already experts with JS / DOM-APIs. Rather than be a high-level explainer (like the last one was) — it goes through a series of examples (each building off the last) to introduce the key problems that template engine has to solve.

**Review by klebba — DISMISSED** [review]

File: doc/TEMPLATE_ENGINE.md [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+Web Components. That’s because from the perspective of markup… custom elements
+_are just plain html_! So, there’s nothing _special_ a template engine really
+needs to do with respect to custom elements.
+
+## What is an html template engine, anyways?
+
+An html template engine allows developers to ergonomically create dynamic web
+applications — plain and simple.
+
+One way to accomplish this would be to create an entire html document in a
```
Comment (klebba) [inline doc/TEMPLATE_ENGINE.md:22]: IMO the server side example could be omitted here

File: doc/TEMPLATE_ENGINE.md [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+
+// Pick a container to render DOM in.
+const container = document.body;
+
+// Set up our application state.
+const data = {
+  label: 'My TODOs',
+  selected: 0,
+  items: [
+    { label: 'walk the dog', description: '…' },
```
Comment (klebba) [inline doc/TEMPLATE_ENGINE.md:59]: nit — use `...` in code vs ellipsis symbol

File: doc/TEMPLATE_ENGINE.md [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+Moving forward, we’ll just call it a “template engine”.
+
+## So, how do you go about building one?
+
+Well… let’s try it! In general, there are three concepts to consider: template
+definition, data hydration, and result rendering. In particular, the desired
+interface ought to be something like:
+
+```js
+render(container, html`<div>markup</div> + <div>${data}</div>`);
```
Comment (klebba) [inline doc/TEMPLATE_ENGINE.md:40]: nit: the `+` looks like an operator

File: doc/TEMPLATE_ENGINE.md [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+  ],
+};
+
+// Derive rendering-specific information from our state.
+const label = data.label;
+const items = data.items;
+const selectedItem = data.items[data.selected];
+
+// Interpolate our template with our state idiomatically.
+const result = html`
```
Comment (klebba) [inline doc/TEMPLATE_ENGINE.md:71]: > Following is an example which contains: 
> an attribute binding on line
> a property binding on line
> a list on line
> nesting within the list
> a conditional statement

These cover most of the cases the engine must handle

File: doc/TEMPLATE_ENGINE.md [hunk 5]
```diff
@@ -0,0 +1,619 @@
...
+`;
+
+// Finally, we perform the update.
+render(container, result);
+```
+
+### Iteration 1 — a first crack
+
+```js
+// Recursively create html by joining strings with values.
```
Comment (klebba) [inline doc/TEMPLATE_ENGINE.md:104]: nit: densify the comments here

File: doc/TEMPLATE_ENGINE.md [hunk 6]
```diff
@@ -0,0 +1,619 @@
...
+* Hydration is idiomatic JS.
+* It’s composable — i.e., you can nest, loop, etc.
+* Content can be interpolated between tags (e.g., text in a paragraph).
+* Attributes can be interpolated inside opening tags (e.g., flag on an item).
+* **It understands html.**
+* **Boolean attributes and properties work via `.` / `?` DSL.**
+
+**What it doesn’t do**
+
+* Nothing is cached and the DOM is thrashed.
```
Comment (klebba) [inline doc/TEMPLATE_ENGINE.md:335]: elaborate on `thrashed`

File: doc/TEMPLATE_ENGINE.md [hunk 7]
```diff
@@ -0,0 +1,619 @@
...
+* It understands html.
+* Boolean attributes and properties work via `.` / `?` DSL.
+* **DOM and HTML is cached (limited thrashing).** †
+* **All scripted implementation details are internalized.**
+* **Change-by-value detection.**
+
+**What it doesn’t do**
+
+* Some markup abstractions are leaking.
+* Many edge cases are not considered.
```
Comment (klebba) [inline doc/TEMPLATE_ENGINE.md:603]: nit: `e.g.:`

File: doc/TEMPLATE_ENGINE.md [hunk 8]
```diff
@@ -0,0 +1,619 @@
...
+  `strings` argument will _always be exactly the same_. That means you can stick
+  it in a weak map.
+
+## Wrapping up!
+
+In just a couple iterations — we went from a naive implementation which just
+does recursive string concatenation… all the way to an html-aware, performant
+DOM management system. Nice!
+
+The remaining edge-cases and markup-cleanup won’t be covered here, but this
```
Comment (klebba) [inline doc/TEMPLATE_ENGINE.md:617]: perhaps "wont be covered here, but from here we would direct you to the source itself"

**Comment by theengineear** [comment]
Iterated on nearly all the comments you left in here @klebba — the one main thing that I am going to put to later is adding additional _code comments_ to the example code. I think that will actually take quite a bit more time to thoughtfully do and I’d like to pause on this for now.

Screenshot 2024-10-18 at 5 02 30 PM

**Pull request merged**